pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,20 +283,40 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
-		// make a binding reference to the claim.
-		// triggers update of the claim in this controller, which builds claim status
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim and ensure to update the local index to prevent dupe bindings
+		clone, err := conversion.NewCloner().DeepCopy(volume)
+		if err != nil {
+			return fmt.Errorf("Error cloning pv: %v", err)
+		}
+		volumeClone, ok := clone.(*api.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+		}
+		volumeClone.Spec.ClaimRef = claimRef
+		if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeClone); err != nil {
+			return fmt.Errorf("Unexpected error saving PersistentVolume.Status: %+v", err)
+		} else {
+			volumeIndex.Update(updatedVolume)
```

Comment:
If I'm reading correctly, this is only valid because the `binder.lock` associated with the `binder.volumeIndex` is held when this is called.  That's worthy of the comment on the method.